java-checks/src/main/java/org/sonar/java/checks/regex/UnicodeAwareCharClassesCheck.java

```diff
@@ -0,0 +1,110 @@
+  private static final Set<String> unicodeAwarePropertiesWithFlag = ImmutableSet.of(
+    "Lower", "Upper", "Alpha", "Alnum", "Punct", "Graph", "Print", "Blank", "Space");
+
+  private static final Map<Character, Character> unicodeUnawareCharacterRanges = ImmutableMap.of(
+    'a', 'z', 'A', 'Z');
+
+  @Override
+  public void checkRegex(RegexParseResult regexForLiterals, MethodInvocationTree mit) {
+    UnicodeUnawareCharClassFinder unicodeUnawareCharClassFinder = new UnicodeUnawareCharClassFinder();
+    unicodeUnawareCharClassFinder.visit(regexForLiterals);
+
+    unicodeUnawareCharClassFinder.unicodeUnawareRange.forEach(unawareRange -> reportIssue(unawareRange,
+      "Replace this character range with a Unicode-aware character class.", null, Collections.emptyList()));
+
+    if (!unicodeUnawareCharClassFinder.unicodeAwareWithFlag.isEmpty() &&
+      !unicodeUnawareCharClassFinder.containsUnicodeCharacterFlag()) {
+
+      List<RegexCheck.RegexIssueLocation> secondaries = unicodeUnawareCharClassFinder.unicodeAwareWithFlag.stream()
+        .map(tree -> new RegexIssueLocation(tree, "Predefined/POSIX character class"))
+        .collect(Collectors.toList());
+      reportIssue(ExpressionUtils.methodName(mit),
+        "Enable the \"UNICODE_CHARACTER_CLASS\" flag or use a Unicode-aware alternative.", null, secondaries);
+    }
+  }
+
+  private static class UnicodeUnawareCharClassFinder extends RegexBaseVisitor {
+
+    private final List<RegexTree> unicodeUnawareRange = new ArrayList<>();
+    private final List<RegexTree> unicodeAwareWithFlag = new ArrayList<>();
+    private boolean containsUnicodeCharacterFlag = false;
+
+    @Override
+    public void visitCharacterRange(CharacterRangeTree tree) {
+      Character expectedUpperBoundChar = unicodeUnawareCharacterRanges.get(tree.getLowerBound().getCharacter());
+      if (expectedUpperBoundChar != null && expectedUpperBoundChar == tree.getUpperBound().getCharacter()) {
+        unicodeUnawareRange.add(tree);
+      }
+    }
+
+    @Override
+    public void visitEscapedCharacterClass(EscapedCharacterClassTree tree) {
+      String property = tree.property();
+      if ((property != null && unicodeAwarePropertiesWithFlag.contains(property)) ||
+        unicodeAwareClassesWithFlag.contains(tree.getType())) {
+
+        unicodeAwareWithFlag.add(tree);
+      }
+    }
+
+    @Override
+    protected void before(RegexParseResult regexParseResult) {
+      containsUnicodeCharacterFlag |= regexParseResult.getInitialFlags().contains(Pattern.UNICODE_CHARACTER_CLASS);
+    }
+
+    @Override
+    protected void doVisitNonCapturingGroup(NonCapturingGroupTree tree) {
+      containsUnicodeCharacterFlag |= flagActive(Pattern.UNICODE_CHARACTER_CLASS);
+      super.doVisitNonCapturingGroup(tree);
+    }
+
+    public boolean containsUnicodeCharacterFlag() {
+      return containsUnicodeCharacterFlag;
+    }
```

Comment:
I think this getter can be removed. You access the other variables without a getter as well and if you move the code using the variables into `after`, there's no need for getters anyway.